// preview-email/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

const fs = require('fs');

const os = require('os');

const path = require('path');

const dayjs = require('dayjs');

const debug = require('debug')('preview-email');

const nodemailer = require('nodemailer');

const open = require('open');

const pify = require('pify');

const pug = require('pug');

const uuid = require('uuid');

const {
  simpleParser
} = require('mailparser');

const writeFile = pify(fs.writeFile);
const transport = nodemailer.createTransport({
  streamTransport: true,
  buffer: true
});
const templateFilePath = path.join(__dirname, '..', 'template.pug');
const renderFilePromise = pify(pug.renderFile);

const previewEmail = async (message, options) => {
  options = _objectSpread({
    dir: os.tmpdir(),
    id: uuid.v4(),
    open: {
      wait: false
    },
    template: templateFilePath
  }, options);
  debug('message', message, 'options', options);
  if (typeof message !== 'object') throw new Error('Message argument is required');
  const res = await transport.sendMail(message);
  const parsed = await simpleParser(res.message);
  console.log('parsed', parsed);
  console.log('parsed.attachments', parsed.attachments);
  const html = await renderFilePromise(options.template, Object.assign(parsed, {
    cache: true,
    pretty: true,
    dayjs
  }));
  const filePath = `${options.dir}/${options.id}.html`;
  debug('filePath', filePath);
  await writeFile(filePath, html);
  if (options.open) await open(filePath, options.open);
  return `file://${filePath}`;
};

module.exports = previewEmail;
